course-material/intro-to-sql/aggregate-functions-in-sql.md

```diff
@@ -1,14 +1,287 @@
+
+```{code-cell} ipython3
+%%sql 
+SELECT MIN(balance) AS 'Minimum Balance'
+FROM bank 
+```
+
+```{code-cell} ipython3
+%%sql 
+SELECT MAX(balance) AS 'Maximum Balance'
+FROM bank 
+```
+
+Accumulating everything we have learned so far, can you think of another way of finding the minimum and maximum of balance without `MIN()` and `MAX()`? <b>Hint:</b> try recreating these queries with `ORDER BY`.
+
+#### Grouping
+
+Grouping is an extremely useful clause. It allows users to examine the results of aggregate functions within each unique group. Note that grouping with `GROUP BY` comes after filtering with `WHERE`. Below, we find the `COUNT()` of all rows `GROUP BY` "housing". Since housing only has the unique values of 'yes' and 'no', there will be only two groups.  
+
+```{code-cell} ipython3
+%%sql 
+SELECT COUNT(*) AS Count
+FROM bank 
+GROUP BY housing
+```
+
+The problem with the above query is we can not determine what value belongs to which group. To fix this, `SELECT` both the "housing" variable and the `COUNT(*)` function. Let's also change 'housing' to 'Housing' in our final output just for demonstration purposes.
+
+```{code-cell} ipython3
+%%sql 
+SELECT housing as Housing, COUNT(*) AS Count
+FROM bank 
+GROUP BY housing
+```
+
+We can now clearly see which count belongs to which group under "housing". 
+
+`GROUP BY` also allows for grouping with several variables. For instance, let's first `GROUP BY` "housing" and then `GROUP BY` "marital". Then, find the `COUNT()` of these groups.
+
+```{code-cell} ipython3
+%%sql 
+SELECT housing AS Housing, marital AS Marital, COUNT(*) AS Count
+FROM bank 
+GROUP BY housing, marital
+```
+
+There are six total groups from our query. "housing" has two groups and "marital" has three groups. Since "housing" has only two groups, 'yes' and 'no', let's think about these groups as two seperate blocks. The 'yes' and 'no' blocks will then each be `GROUP BY` the groups under the "marital" column, which are 'married', 'single', and 'divorced'. Thus, since each 'yes' and 'no' group has three groups each, there are a total of six groups.
+
+[![](https://mermaid.ink/img/pako:eNqVkjFvgzAQhf-KdV1JBAc44EodqnRMF7q0pYOF3cQK2Mg4bWiU_14H1EhJlrL5fe_5nWTfASojJDBYW95uyMtjqQl5fS-hl10JH4Mis9kDWUUeNtxaJcWFUZyMTul1LS_48sSF-jK2Ot949kibPzH24k3vaBR43TvyJd721kpvC9fXkoSkc9ZsJbur8ozzKBjl7FsJt2HY7u8v89HEPE7MxxPzycR8-o88BNBI23Al_D8fTvdLcBvZ-Hdl_ii43ZZQ6qPP8Z0zRa8rYM7uZAC7VnAnl4r79WiAffK6O9MnoZyxZygHuRq3aViqAFqugR1gDyxK8znGaRoj0miBOWIAvceYzGm2SDDNKGKKeAzgxxg_KpxnmGVJhGEe0gWlMR3q3gZzGHn8BXRf3tk?type=png)](https://mermaid.live/edit#pako:eNqVkjFvgzAQhf-KdV1JBAc44EodqnRMF7q0pYOF3cQK2Mg4bWiU_14H1EhJlrL5fe_5nWTfASojJDBYW95uyMtjqQl5fS-hl10JH4Mis9kDWUUeNtxaJcWFUZyMTul1LS_48sSF-jK2Ot949kibPzH24k3vaBR43TvyJd721kpvC9fXkoSkc9ZsJbur8ozzKBjl7FsJt2HY7u8v89HEPE7MxxPzycR8-o88BNBI23Al_D8fTvdLcBvZ-Hdl_ii43ZZQ6qPP8Z0zRa8rYM7uZAC7VnAnl4r79WiAffK6O9MnoZyxZygHuRq3aViqAFqugR1gDyxK8znGaRoj0miBOWIAvceYzGm2SDDNKGKKeAzgxxg_KpxnmGVJhGEe0gWlMR3q3gZzGHn8BXRf3tk)
```

Comment:
save this image locally and add it to the repo, link might be taken down

share image source